### ISSUE TITLE
Критическая уязвимость RCE

### ISSUE THREAD

**Issue by iliakonnov** [stasivs/Messenger#7]
Есть возможность выполнять произвольный код на стороне сервера, отправив специально сформированное сообщение. PoC: https://github.com/iliakonnov/PickleMessengerExploit
Потенциально можно развить уязвимость и таким же образом атаковать всех подключенных к серверу клиентов.

**Comment by brinza888** [comment]
Интересно. Посмотрю exploit и займусь устранением этой дыры. Думаю, нужна грамотная обратка приходящих данных.

**Comment by brinza888** [comment]
Очевидно, что в Messages.py, в классе UserMessage, в методе get_message() нужно изменить форматный вывод через .format(), на шаблонный через String.Template. Тогда мы избежим исполнения кода, который был отправлен пользователем в качестве сообщения.

**Comment by iliakonnov** [comment]
Нет, проблема совсем не в string.format. Проблема в самом pickle. Между прочим, на странице документации по нему имеется большой красный warning с описанием. https://docs.python.org/3/library/pickle.html
Если интересны подробности реализации PoC, то предлагаю списаться через вк. Гитхаб не очень удобный

**Comment by iliakonnov** [comment]
Подключение и вывод в консоль сервера произвольного текста:
![messenger](https://user-images.githubusercontent.com/9402179/52990221-3f74f600-3418-11e9-82f6-b9cff71e6fae.gif)

Смена прав по-умолчанию:
![3xv00g3a0lc](https://user-images.githubusercontent.com/9402179/52989831-3be06f80-3416-11e9-8f1d-2b42fefae202.jpg)

Выполнение на сервере произвольного кода:
![fwbwtg84qrw](https://user-images.githubusercontent.com/9402179/52989860-629ea600-3416-11e9-8604-b51c2eff04f4.jpg)

Выполнение команд с повышенными привилегиями (прям как sudo, но без пароля):
![qkwu-riyygg](https://user-images.githubusercontent.com/9402179/52989880-819d3800-3416-11e9-8348-d03b2eaf3dee.jpg)

**Comment by brinza888** [comment]
Давай спишемся в vk.
Вот мой: vk.com/brinza888
